implement choose subcommand and support use as macOS default browser via platypus

mtoohey31 opened the pull request:
Hello! As the title mentions, this PR implements a new subcommand that prompts a user to select which profile they'd like to use for the provided command, url, etc., using rofi, dmenu, or applescript on macOS. It can also be run without any additional qutebrowser arguments to open a new window from that profile. It also modifies the desktop subcommand a little bit to allow for the creation of a desktop file by the name of `qbpm.desktop` that simply runs this choose command. In order to support this, I had to modify the `desktop` subcommand a little to support a variable number of profiles to create desktops for, so that it could be run as just `qbpm desktop -c` without specifying profiles. I hope that's ok, if there's different approach to this you'd prefer, let me know and I can rework it. The other thing I was looking for input on with this change is whether this file should automatically be created, say, after more than 1 profile has been created. Let me know what your opinion is on that!

The second commit makes use of [platypus](https://github.com/sveinbjornt/Platypus) to support the use of this `choose` subcommand as the macOS default browser; I included them in the same PR because the second relies on the first. I've added installation instructions to the README, one caveat with this PR though is that there's currently a bug in qutebrowser that results in two tabs being opened, one with the actual url, and one with the format `file:///{url}`. Unfortunately I haven't found a workaround yet, but I've mentioned the bug in the README under the macOS installation instructions, so users understand where the issue is coming from. The one other thing I should mention is the creation of the `qbpm-choose` script in the repo is unfortunately necessary because platypus needs to be pointed to a specific script file.

I think that's pretty much it, lmk if you've got any questions or suggestions about things I should tweak though!

pvsr commented:
This is super cool, thanks! I don't have time to give it a review right now but I'm going on vacation next week so I should be able to get to it then, if not sooner.

mtoohey31 commented:
Thank you, and no worries!

pvsr reviewed (changes requested):
Sorry for the delay here, I've been enjoying some offline time during my vacation.

> It also modifies the desktop subcommand a little bit to allow for the creation of a desktop file by the name of qbpm.desktop that simply runs this choose command. In order to support this, I had to modify the desktop subcommand a little to support a variable number of profiles to create desktops for, so that it could be run as just qbpm desktop -c without specifying profiles. I hope that's ok, if there's different approach to this you'd prefer, let me know and I can rework it.

I would prefer to have a static `qbpm.desktop` file in the repo, and leave the desktop subcommand unchanged. Does that work for you?

mtoohey31 commented on qpm/utils.py:
```diff
@@ -26,3 +29,18 @@ def user_data_dir() -> Path:
+    check_cmd = subprocess.Popen(["which", command], stdout=subprocess.DEVNULL,
+                                 stderr=subprocess.DEVNULL)
+    check_cmd.communicate()
+    return check_cmd.returncode == 0
```
Missed that, resolved by a43765ee

mtoohey31 commented on qbpm-choose:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+qbpm choose "$@"
```
Good idea, resolved by 349212b1

mtoohey31 commented:
> Sorry for the delay here, I've been enjoying some offline time during my vacation.

No worries :slightly_smiling_face:

> I would prefer to have a static `qbpm.desktop` file in the repo, and leave the desktop subcommand unchanged. Does that work for you?

Sure, that seems like a cleaner solution! Should I add manual installation instructions for that to the README, or should I look into finding a way for it to be installed automatically with `pip install`, not sure if that's possible?

mtoohey31 commented on qbpm.platypus:
```diff
@@ -0,0 +1,259 @@
+	<key>Overwrite</key>
+	<false/>
+	<key>PromptForFileOnLaunch</key>
+	<true/>
```
You're right, that seems to have been a mistake, that behaviour wasn't intended (fixed with 32539f4d). I'm somewhat confused by the alternative behaviour as well though: I would expect it to run `qbpm choose` with no arguments, which from the command line results in a choice between profiles to open a new window for, but instead, it just does nothing and the application closes immediately. I'm not really sure what's going on, lmk if you've got any ideas, but otherwise that behaviour is better than the file picker.

mtoohey31 commented:
Ok so I reverted the changes to the desktop subcommand and added the static file as you suggested. I think the only thing left to consider is the strange behaviour I mention [here](https://github.com/pvsr/qbpm/pull/21#discussion_r665767057). Again though, I've kind of reached a dead end there, and it's not necessarily buggy behaviour, just a little unexpected.

pvsr commented:
Okay this pretty much looks good so I'm going to merge and fix up anything left over. Thanks for your contribution!

> Sure, that seems like a cleaner solution! Should I add manual installation instructions for that to the README, or should I look into finding a way for it to be installed automatically with `pip install`, not sure if that's possible?

Yeah as far as I know it's not possible to install through pip unfortunately. I'll mention it in the readme.

> I think the only thing left to consider is the strange behaviour I mention [here](https://github.com/pvsr/qbpm/pull/21#discussion_r665767057). Again though, I've kind of reached a dead end there, and it's not necessarily buggy behaviour, just a little unexpected.

After some investigation, I think the issue may be that when macOS starts applications it gives them a limited $PATH, just /bin:/usr/bin/:/sbin:/usr/sbin, and end user programs like qbpm are usually not installed in any of those dirs. That can be fixed by running `bash -l` as the interpreter so the full user $PATH gets set.

mtoohey31 commented:
> After some investigation, I think the issue may be that when macOS starts applications it gives them a limited $PATH, just /bin:/usr/bin/:/sbin:/usr/sbin, and end user programs like qbpm are usually not installed in any of those dirs. That can be fixed by running `bash -l` as the interpreter so the full user $PATH gets set.

Ah, that would explain it, thanks for figuring that out, and thanks for the pointers!
